### ISSUE TITLE
Misc : Login Crawler

### ISSUE THREAD

**Issue by haizad** [Leantime/leantime#296]
Pardon my action for creating this issue, but I need to find a way on how to integrate Leantime with my existing project.
I am using Keycloak as a authentication / user management. Due to limitation on sso at this moment, I can't find a way to merge my project with Leantime. What comes into my mind is making login through cURL. like this one

$ch = curl_init();

curl_setopt($ch, CURLOPT_URL, 'http://xxxxx');
curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
curl_setopt($ch, CURLOPT_CUSTOMREQUEST, 'POST');
curl_setopt($ch, CURLOPT_POSTFIELDS, "redirectUrl=/&username=xxxx&password=xxxx&login=Login");
curl_setopt($ch, CURLOPT_ENCODING, 'gzip, deflate');

$headers = array();
$headers[] = 'Connection: keep-alive';
$headers[] = 'Pragma: no-cache';
$headers[] = 'Cache-Control: no-cache';
$headers[] = 'Upgrade-Insecure-Requests: 1';
$headers[] = 'Origin: http://xxxxx';
$headers[] = 'Content-Type: application/x-www-form-urlencoded';
$headers[] = 'User-Agent: Mozilla/5.0 (iPhone; CPU iPhone OS 10_3_1 like Mac OS X) AppleWebKit/603.1.30 (KHTML, like Gecko) Version/10.0 Mobi$
$headers[] = 'Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;$
$headers[] = 'Referer: http://xxxxx';
$headers[] = 'Accept-Language: en-US,en;q=0.9';
$headers[] = 'Cookie: lang=en-US; sid=xxxxxxxxx-xxxxxxxx';
curl_setopt($ch, CURLOPT_HTTPHEADER, $headers);

$result = curl_exec($ch);
if (curl_errno($ch)) {
    echo 'Error:' . curl_error($ch);
}

I found out it relates with sid cookies, so I create one using this code

$password = 'userpass';
$tmp = hash('sha1', (string)mt_rand(32,32) . $_SERVER['REMOTE_ADDR'] . time());
// echo $_SERVER['REMOTE_ADDR'];
$sid=$tmp .'-'.hash('sha1', $tmp.$password);

But the sid does not work that way. Any Idea on how to make it work?

I also tried to use Goutte (PHP Web Scraper) but it only returns me the html code of the dashboard.

I am sorry for digging your codes of work, but I am kind of desperate somehow. :/

**Comment by marcelfolaron** [comment]
Interesting approach. My assumption would be that the user wouldn't receive the cookie as you have already discovered. The problem with setting it manually is that the session id in the database would be different.

You would need to check the cookie coming from your curl request (https://stackoverflow.com/questions/895786/how-to-get-the-cookies-from-a-php-curl-into-a-variable) and then set it manually using curl (https://stackoverflow.com/questions/16872082/how-can-i-send-cookies-using-php-curl-in-addition-to-curlopt-cookiefile)

Let me know if that works. I'd be interested to see the results

**Comment by haizad** [comment]
Thank you for your response, and sorry for the late reply. 

I managed to receive the sid cookie, and set it back manually back using curl, but I guess there is more than that need to be configured, like sessions , csrf.. etc. 

The sid that were received does not work even if I get sid from the database , or update a new one into the database first.

The interesting part is I can managed to log in back by using the sid that I got right after I logged in as usual on the browser.

![Screenshot from 2020-06-26 07-32-08](https://user-images.githubusercontent.com/23258127/85805723-ada19980-b77f-11ea-9cd1-f2c6cd4130e8.png)

I guess it relates with other sessions, that I didn't take care of.

:)
